lib/internal/modules/esm/loader.js

```diff
@@ -100,6 +100,26 @@ const {
 
 let emittedSpecifierResolutionWarning = false;
 
+/**
+ * A utility function to iterate through a hook chain, track advancement in the
+ * chain, and generate and supply the `next<HookName>` argument to the custom
+ * hook.
+ * @param {KeyedHook[]} chain The whole hook chain.
+ * @param {object} meta Properties that change as the current hook advances
+ * along the chain.
+ * @param {boolean} meta.chainFinished Whether the end of the chain has been
+ * reached AND invoked.
+ * @param {string} meta.hookErrIdentifier A user-facing identifier to help
+ *  pinpoint where an error occurred. Ex "file:///foo.mjs 'resolve'".
+ * @param {number} meta.hookIndex A non-negative integer tracking the current
+ * position in the hook chain.
+ * @param {string} meta.hookName The kind of hook the chain is (ex 'resolve')
+ * @param {boolean} meta.shortCircuited Whether a hook signaled a short-circuit.
+ * @param {function(hookErrIdentifier, hookArgs)} validate A wrapper function
+ *  containing all validation of a custom loader hook's intermediary output. Any
+ *  validation within MUST throw.
```

Comment:
You can use the TypeScript form to include the return type of the function:

```suggestion
 * @param {(hookErrIdentifier, hookArgs) => void} validate A wrapper function
 *  containing all validation of a custom loader hook's intermediary output. Any
 *  validation within MUST throw.
```

(I assume this function either throws or returns nothing/void?)

Also we have the longer line lengths now, if you want to wrap these any farther out.

Reply:
Yes, it only throws or doesn't. I think that's `=> never`?

Reply:
I’ve never heard of `never`. If you’re using VS Code, you can hover over `validate` on the declaration line and the tooltip will tell you what the TypeScript server recognizes your JSDoc comment as.

Reply:
`never` is how you indicate in TS that it always throws.

Reply:
I assume it only throws on error, right, not always? So when it doesn't throw, it returns void.